### ISSUE TITLE
NODE_ENV test affects Projen behavior

### ISSUE THREAD

**Issue by hs-kgoriunov** [projen/projen#2947]
Projen has an implicit behavior of disabling logging and synthesizing project files to a tmp directory when it [detects](https://github.com/projen/projen/blob/454794ab44889e965dbbde87a51951c86269e3f8/src/common.ts#L5) the `NODE_ENV=test` environment variable. Due to a legacy decision, I have a CI workflow that runs Projen from within a Docker image that has the environment variable set globally.

It would be nice, to have production builds of Projen not to respect the environment variable. Otherwise, it's difficult to provide a different environment solely to Projen. For example, when I run `yarn ci` I'm actually calling `npx projen ci`, which in turn calls all sort of intermediate project-dependent scripts. Some of the scripts require NODE_ENV to be test, while Projen itself wants NODE_ENV to be something else.

This place, in particular, makes Projen to keep silence in CI logs about what it does, which is inconvenient when something doesn't work as expected.
https://github.com/projen/projen/blob/a003bf3707c176083444a950630f438ea2cd75ac/src/logging.ts#L6-L15

---

On the other side, I understand that such change can possibly break other libraries that use Projen not as a tool, but as a framework to build their own project synthesizers. So it might make sense to disrespect the test environment only when Projen is called via its `bin/projen` entry script.

**Comment by mrgrain** [comment]
Yikes, this is bad. I think the only reason we have this is for the tests in projen itself.
For which we absolutely do not have to use `NODE_ENV`.

**Comment by mrgrain** [comment]
@hs-kgoriunov Apologies, but the change has to be reverted. This unintentionally broke the documented approach to testing custom project types (https://projen.io/docs/testing/#common-issues-and-solutions) 
Bug report: https://github.com/projen/projen/discussions/3322#discussion-6178633

`NODE_ENV` is the only way to detect of projen code is runing as part of a test or not. I'm not inclined to say this is a won't fix.

However we might be able to target the testing behavior better with additional env variables?

Alternatively, I think this might solve your specific problem:
```
NODE_ENV=prod npx projen ci
````

Also you might be able to add this to specific commands.

**Comment by hs-kgoriunov** [comment]
> Alternatively, I think this might solve your specific problem

I can't do that, the `ci` command won't use the correct configuration if something apart from `test` is used for the node env.

I already solving it by selectively changing the behavior of the affected Projen tasks with the `env: { NODE_ENV: "projen" }`, while keeping everything else to the `test` environment. It just feels weird that Projen needs this kind of special treatment.

---

To my understanding, we have two Projen flavors—Projen as a tool and Projen as a framework. The first one shouldn't get impacted by the environment, while the second should. Hence, the contradiction.

**Comment by mrgrain** [comment]
> To my understanding, there are two Projen flavors—Projen as a tool and Projen as a framework. The first one shouldn't get impacted by the environment, while the second should. Hence, the contradiction.

I'm not sure where **you** draw the line between tool and framework. Most of the code, including logging, is wrapped up inside project construction and `synth()`. 

The [documented approach to testing](https://projen.io/docs/testing/#testing) should keep working, otherwise it's a breaking change. While we can make these, we also need an alternative and update docs. I'm not confident we have a viable alternative at the moment. Definitely open to suggestions.

**Comment by mrgrain** [comment]
But I think your point is the following:

- `npx projen ci` defines a task `yarn ci`
- projen doesn't do anything but passing through to the actual command
- And that output is currently ignores, but shouldn't?

Is that a fair assumption?

**Comment by hs-kgoriunov** [comment]
> I'm not sure where you draw the line between tool and framework

In my terminology, a tool is like a production build of an application that can have all the tests ripped off without affecting the application. For frameworks, it is different, it's when you're reusing parts of the existing application's codebase to make your own customized version of it. In that case, while the application is still in the development, you'd benefit from having the ability to test it. But when you're ready to build your variation for production, you should be able to rip the reliance on tests and testing environment.

In terms of pseudocode:
```js
Projen {
  // should be detected somehow
  // without the relience on the environment variables
  // maybe getting set via a build scripts, idk
  isProduction = false;  

  synth() {
    shouldUseTempDir = process.env.NODE_ENV == "test" && !Projen.isProduction
  }

  log() {
    shouldLog = process.env.NODE_ENV == "test" && !Projen.isProduction
  }
}
```

---

> Is that a correct interpretation?

`ci` is a multistep Projen task that executes commands and spawns other Projen tasks, which also execute commands. Most of the commands require the test environment, while Projen itself disables logging because of the environment. It's inconvenient to trace execution failures because, well, there are no logs.

**Comment by mrgrain** [comment]
Thanks for the detailed explanation. Your pseudo code seems to imply that we should have a production build version that has this stuff disabled (or ripped out). To be clear, the reason we needed to revert is _not_ because projen needs to test itself. That is a concern that the original attempt solved without issues. The problem is that the production build of projen has a feature that makes testing easier. The canonical way to detect (and disable) that is `NODE_ENV=test`. Like I've said before, I agree that this is somewhat problematic. But I'm equally unwilling to remove this feature.

To summarize, in my opinion we have the following options:
- (1) Break the feature and put the work on the users to make tests work alright. _This is very unlikely to happen_
- (2) Break the feature and find a different way easy way for writing tests using projen
- (3) Keep the feature as is, but provide ways to change with the behavior that projen exhibits, i.e. re-enable logging and outdir

Reading between the lines I am guessing that (3) would not satisfy your use case.
